### ISSUE TITLE
Weather plugin problem

### ISSUE THREAD

**Issue by spyjo** [OpenJabNab/OpenJabNab#47]
Hello,

I have setup Openjabnab this week end, nabaztag is connected correctly to the server.
I try to make the weather plugin work, I have added the city, make it default, and added the previmeteo token. So I think everything is setup ok.

But when I call the weather app it does nothing.
Here is the dump.log:

17/05/2016 19:54:35 - Violet Api Call - /ojn/FR/api.jsp?sn=xxxxxxx&token=yyyyyyy&plugin=weather&function=get
17/05/2016 19:54:35 - Violet Api Answer - <?xml version="1.0" encoding="UTF-8"?><rsp></rsp>

Any idea to make it work ?

Thanks

**Comment by spyjo** [comment]
Just to make an update on this problem:
- if webcast cron is setup, the weather is spoken correcly at the cron time
- if called by the api url, it does nothing:
  `22/05/2016 10:07:58 - Violet Api Call - /ojn/FR/api.jsp?sn=xxxxxxx&token= yyyyyyy&plugin=weather&function=get`
  `22/05/2016 10:07:58 - Violet Api Answer - <?xml version="1.0" encoding="UTF-8"?><rsp></rsp>`
- if called by the onclick button, it just crash the server. Here is the log before it crash:
  `22/05/2016 10:14:06 - XMPP Bunny (xxxxxxx) - <message from='xxxxxxx@my.domain.com/idle' to='int@xmpp.objects.violet.net/int' id='12'><button xmlns="violet:nabaztag:button"><clic>1</clic></button></message>
  `

Any advice ? Thanks

**Comment by RedoXyde** [comment]
Hi,

I just set up a new server for development, and haven't been able to reproduce your issues:
- Cron job works fine (as you mentionned)
- Single click works fine with the default city.  Make sure you have selected the Weather plugin as "Single click plugin" on the main config page for the Nabaztag. Also, make sure you have selected a default city in the plugin configuration. 
- RFID/ZTamp association works fine

Also, there is not "get" apicall for this plugin, according to the "source" ( see: [plugin_weather.cpp:118](https://github.com/OpenJabNab/OpenJabNab/blob/master/server/plugins/weather/plugin_weather.cpp#L118) ), so I guess it's completely normal that it doesn't work.

It would be a big help if you could recompile the server in Debug configuration:
`make clean
./toDebug.sh
make`
And then run the server in gdb:
`gdb -ex run ./openjabnab`
When a crash append: type `bt` and just copy/paste the backtrace here (then you can restart the server using `r` or exit the debugger `quit`). If you can't or don't want to, it's no big deal but I don't think we can do much more, since I can't reproduce the issue.

Best,
-RedoXyde

**Comment by spyjo** [comment]
Yes I did setup the weather plugin as single click, and the default city is configured.

I have recompiled the server in debug mode, restarted, and ... now it's also working with single click button. Strange but great :)
I will keep it running in gdb until I see again this crash.
There is just one error displayed but it does not seems to be a problem since it's working:
"QObject: Cannot create children for a parent that is in a different thread."

For the "get" weather api call, you are right I don't see it either in the source. But this call is working on openjabnab.fr , so I tought the server was the same here. It seems not. Is there a way to call the weather from an API url on the github version ?

Thanks

**Comment by RedoXyde** [comment]
Hi,

Thanks for the feedback. 

I just tried recompiling in Release mode, but still couldn't make it crash.

I also get the same warning, either in Release or Debug mode. I'm not sure it's supposed to happen, but it still works. I'm pretty sure it can be fixed some way or another...

About the "get" APICall @Pixel166 made **a lot** of changes on his server (http://openjabnab.fr) over the years and he didn't merge them back (not enough time and/or too much work). I'd say he's got a different version of the plugin.

Adding the APICall to the current version (the Github one) shouldn't be a problem, but requires a bit of work. You just needs to add a _new_ API call that triggers the same action as the SingleClick plugin does. @drakaz made the last changes on this plugin, maybe he will be okay to help, if you need. I could also do it if no one has time to add this feature (but I don't have much time to work on this project).

Best,
-RedoXyde

**Comment by spyjo** [comment]
I was able to modify the plugin to create the get api, but in fact I still can't use it with the "violet api" which has a static token, it's just working with the http api which need auth to get a temporary token (if I understand well how it works), so I think it's not very usefull for now, not sure..

Here is what I have done:

```
--- a/server/plugins/weather/plugin_weather.cpp
+++ b/server/plugins/weather/plugin_weather.cpp
@@ -131,6 +131,22 @@ void PluginWeather::InitApiCalls()
        DECLARE_PLUGIN_BUNNY_API_CALL("getlang()", PluginWeather, Api_getLang);
        DECLARE_PLUGIN_BUNNY_API_CALL("setprevitoken(apitoken)", PluginWeather, Api_setPrevitoken);
        DECLARE_PLUGIN_BUNNY_API_CALL("getprevitoken()", PluginWeather, Api_getPrevitoken);
+
+      DECLARE_PLUGIN_BUNNY_API_CALL("get()", PluginWeather, Api_get);
+
+}
+
+PLUGIN_BUNNY_API_CALL(PluginWeather::Api_get){
+        Q_UNUSED(account);
+        Q_UNUSED(hRequest);
+
+        QString city = bunny->GetPluginSetting(GetName(), "Default/City", "").toString();
+        if(city != "") {
+                getWeatherPage(bunny, city);
+                return new ApiManager::ApiOk(QString("Called get() weather for default city '%1' on bunny '%2'").arg(city, QString(bunny->GetID())));
+        }
+       else
+               return new ApiManager::ApiError(QString("Missing default city for plugin Weather"));
 }

 PLUGIN_BUNNY_API_CALL(PluginWeather::Api_addCity) {




--- a/server/plugins/weather/plugin_weather.h
+++ b/server/plugins/weather/plugin_weather.h
@@ -44,6 +44,7 @@ public:
        PLUGIN_BUNNY_API_CALL(Api_setLang);
        PLUGIN_BUNNY_API_CALL(Api_getPrevitoken);
        PLUGIN_BUNNY_API_CALL(Api_setPrevitoken);
+       PLUGIN_BUNNY_API_CALL(Api_get);

 private:
        void getWeatherPage(Bunny *, QString);

```

**Comment by RedoXyde** [comment]
Hi,

It has been a while since I developped anything related to the API (or the server), but I think, it's completely normal. Let me explain.

When you use this kind of API Call: `/ojn/FR/api.jsp?sn=xxxxxxx&token= yyyyyyy&plugin=weather&function=get` , it's handled by the HTTPhandler, which forwards it to the APIManager that tries to process it.
In the APIManager, you can see it's forwarded to the ProcessBunnyVioletApiCall() function,  which n turns forwards it to the ProcessVioletApiCall() (in bunny.cpp). That's the final handler. If you look at the code, you can see there's no `plugin` or `function` arguments handling.

And that actually makes sense, since it's not _really_ a standard/_Violet API_ call. The plugin system is OpenJabNab specific while the Violet API intended to replace the official API.

The good news is: this feature can easily be added to the ProcessVioletAPICall() function (of bunny.cpp).  I think the using ProcessBunnyApiCall() function (from apimanager.cpp) as an example is a good start, it has all the required functions. Argument parsing should be easier in bunny.cpp (but some tweak of hRequest could be required to properly forward the API Call arguments to the plugin, not sure about that... and it's not required for your particular case: get() since there are no arguments involved.)

The other good news is, you first patch is also required, you didn't do all that for nothing =)

I just had a quick look at the code to figure out why it wasn't working as you expected. I could be completely wrong :/ If you have some spare time, I think it's worth a try. If you don't, I could/will have a look and report then, but I won't have much time before the week-end.

Best,
-RedoXyde

**Comment by spyjo** [comment]
Ok thanks, I will try to implement the missing part in the api processing this week when I have some time
